### ISSUE TITLE
Vulnerability in solr

### ISSUE THREAD

**Issue by philippconzett** [DataverseNO/local.dataverse.no#6]
There has been reported a vulnerability in solr. See the discussion in Teams:

[I går 16:20] Loïc Bergeron
    Due to a vulnerability that was published today regarding the logs in Solr it would be best that no new dataset gest published before the fix.
The fix will most certainly come early next week,
​[I går 16:22] Loïc Bergeron
    the vulnerability in question:
https://www.lunasec.io/docs/blog/log4j-zero-day/ and https://www.cyberkendra.com/2021/12/worst-log4j-rce-zeroday-dropped-on.html. https://github.com/tangxiaofeng7/apache-log4j-poc

​[I går 16:27] Obiajulu Odu
    Solr in Dataverse only generates facets. It has nothing to do with login/authenticating. Dataverse team has not come up with any vulnerability issue.
​[I går 16:30] Loïc Bergeron
    Logs not login and yes it uses log4j2 v 2.14 witch is vulnerable
​[I går 16:31] Obiajulu Odu
    But users don't have access to these logs. It is only the system.
Redigert​[I går 16:32] Loïc Bergeron
    No but published dataset do 
Someone could put a payload inside the name of a file or in a metadata field. That is why I published a warning for the curators.
If it is the wrong chat my apologies, if could you move this warning to a more appropriate location.
Redigert​[I går 16:34] Obiajulu Odu
    Datasets are only published by curators
​[I går 18:04] Roy Dragseth
    System Just a heads up: Blocking all web access outside the uit-web portal is under consideration, that is all traffic to port 80 and 443 on other services will then be blocked until mitigation is implemented.
​[I går 22:36] Roy Dragseth
    System All access to port 80 and 443 from the outside world is now blocked, except for the webpages under https://uit.no

**Comment by philippconzett** [comment]
Here is information from the Dataverse Slack channel on how to fix the issue:

Jim Myers Today at 4:06 AM
My understanding, primarily from @Don Sizemore’s investigation is that the Dataverse software itself isn't (doesn't use a vulnerable version of log4j, but solr does. There is a fix posted in the #ig-devops channel: https://dataversecommunity.slack.com/archives/C01GXN8TYF3/p1639167056027200 .

Don Sizemore
Dear @channel you may have seen news of CVE-2021-44228 going around today. Dataverse itself uses log4j version 1.2 and therefore isn't vulnerable, but Solr is. To protect your solr instance, add the following lines to solr.in.sh:

# enable the following setting to address CVE-2021-44228
SOLR_OPTS="$SOLR_OPTS -Dlog4j2.formatMsgNoLookups=true"

and restart Solr.

**Comment by philippconzett** [comment]
dataverse.no and info.dataverse.no are currently not accessible from networks outside UiT.

In addition to fixing solr.in.sh (see above), is it possible to redirect dataverse.no and info.dataverse.no to this page: [https://uit.no/tavla/artikkel/757753/midlertidig_redusert_tilgang_pa_it-tjenester_te](https://uit.no/tavla/artikkel/757753/midlertidig_redusert_tilgang_pa_it-tjenester_te) ?

**Comment by dragz** [comment]
As soon as anyone can confirm that the mitigation is implemented I can approve unblocking the service.

**Comment by philippconzett** [comment]
Thanks, Roy. That would be good, as we already have been contacted by users from NTNU. I'll ping Obi.

**Comment by Louis-wr** [comment]
The log4j2 vulnerability CVE-2021-44228 as been mitigated on test.dataverse.no and dataverse.no in accordance to :
https://solr.apache.org/news.html#apache-solr-affected-by-apache-log4j-cve-2021-44228

**Comment by dragz** [comment]
Good, I will notify the network techs and allow the service to be unblocked.

**Comment by philippconzett** [comment]
Thanks, all. I'll update the information in Teams and Twitter once dataverse.no (cripslock) and test.dataverse.no (tiffany) are unblocked.

**Comment by dragz** [comment]
dataverse.no is now open again.

**Comment by philippconzett** [comment]
Thanks! I checked, test.dataverse.no is also open.

@dragz: Do you have any information about when info.dataverse.no (a site.uit.no / Wordpress page) will be opened?

**Comment by dragz** [comment]
site.uit.no is hosted externally and isn't blocked. info.dataverse.no seems to be running on a uit-hosted address and is blocked so it doesn't seem to be part of the site.uit.no infrastructure. We will put it on the tasklist, but it might take a while before we can check it.

**Comment by philippconzett** [comment]
info.dataverse.no is run on site.uit.no, the full address is https://site.uit.no/dataverseno/.
info.dataverse.no is only an alias. So, I guess it is only the alias redirect that i hosted on a uit server.

**Comment by dragz** [comment]
Ah, I will put it on the allow list.

**Comment by philippconzett** [comment]
Thanks all!

All DataverseNO services are now accessible again:
- [ ] dataverse.no
- [ ] info.dataverse.no
- [ ] test.dataverse.no

The two TROLLing aliases are also working:
- [ ] trolling.uit.no
- [ ] info.trolling.uit.no

Also, we yesterday evening got the official security alert from Harvard below.

I'm closing this issue.

##### Security alert from Harvard (sent to our TOPdesk address, researchdata@hjelp.uit.no):
Frå: Brooke, Danny <dannybrooke@g.harvard.edu> 
Send: laurdag 11. desember 2021 23:21
Til: Brooke, Danny <danny_brooke@harvard.edu>
Kopi: James Myers <qqmyers@hotmail.com>
Emne: Dataverse Security Advisory - Log4j

Hello Dataverse Admins and Security Contacts,

A critical remote code execution vulnerability in Apache Log4j versions 2.10.0 and greater has been released.  Exploits are publicly available and are actively being used. For more information, you can see any of these sources:

https://nvd.nist.gov/vuln/detail/CVE-2021-44228
https://www.lunasec.io/docs/blog/log4j-zero-day/
https://www.randori.com/blog/cve-2021-44228/

The Dataverse software itself uses log4j version 1.2 and therefore is not suspected to be subject to this vulnerability, but Solr is. To protect your Solr instance, add the following lines to solr.in.sh:

# enable the following setting to address CVE-2021-44228
SOLR_OPTS="$SOLR_OPTS -Dlog4j2.formatMsgNoLookups=true"

and then restart Solr. 

It is critical that you do this as soon as possible. Thank you to Don Sizemore at the Odum Institute for this fix.

For a longer term fix, we are investigating upgrading to log4j 2.15.0 or greater.

Please let us know if you have any questions by contacting us at security@dataverse.org or at support@dataverse.org. You can also reply to this message.

**Comment by philippconzett** [comment]
Harvard has just sent us an email providing additional steps on how to keep our Dataverse installations secure.

@Louis-wr and Obi, please carry out the necessary steps on all our Dataverse installations:
- [ ] dataverse.no (cripslock)
- [ ] test.dataverse.no (tiffany)
- [ ] internal test (henry)

MESSAGE FROM HARVARD:
Brooke, Danny 15 desember 2021, 01:25

Hello Dataverse Admins and Security Contacts,

We previously contacted you about a critical vulnerability in log4j. This is a follow up message and it contains additional steps that you should take in order to keep your installation secure. For more information, you can see one of these sources:

https://logging.apache.org/log4j/2.x/security.html
https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-45046

We do not currently believe that this new vulnerability is exploitable given the standard Dataverse/Solr configuration. However, it is not addressed by updating the solr.in.sh file to include the ‘-Dlog4j2.formatMsgNoLookups=true’ parameter. Given the potential for this new issue to be exploitable, as well as the possibility for additional vulnerabilities to be found that stem from the same underlying functionality, we are now recommending that Dataverse installations remove a class from the log4j jar in solr.

Specifically, use the following commands to remove the JndiLookup.class file from the log4j-core-*.jar file in solr’s server/lib/ext directory. (The version of log4j may vary with solr version - the use of * in the commands should work for all versions of solr.)

From your solr installation directory:

cd server/lib/ext
zip -q -d log4j-core-*.jar org/apache/logging/log4j/core/lookup/JndiLookup.class
service solr restart

To verify, use

unzip -l log4j-core-*.jar | grep JndiLookup.class

and check to make sure JndiLookup.class is no longer found.

It is recommended that you do this as soon as practical. The most recent information indicates that a Denial of Service (DOS) would be the expected worst outcome of a possible exploit, and we are suggesting this out of an abundance of caution.

Thank you to several members of the Dataverse Community who have investigated and discussed the solution above.

For a longer term fix, we are investigating upgrading to log4j 2.16.0 or greater and also upgrading Solr to a soon-to-be-released version that will contain additional fixes.

Please let us know if you have any questions by contacting us at security@dataverse.org or at support@dataverse.org.

**Comment by Louis-wr** [comment]
regarding CVE-2021-45046,
mitigation implemented.

**Comment by philippconzett** [comment]
Great, thanks!
